### PR TITLE
Hardhat rpc

### PR DESCRIPTION
**Pull request by gmbronco**
# Description

Adding basic support for running the app with local test node.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## How should this be tested?

To start a node run: `npm run node` script is starting a node with a pinned block.

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have requested at least 2 reviews (If the PR is significant enough, use best judgement here)
- [ ] I have commented my code where relevant, particularly in hard-to-understand areas
- [x] If package-lock.json has changes, it was intentional.
- [x] The base of this PR is `master` if hotfix, `develop` if not

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202244598728459

### CODE REVIEW

**Review by garethfuller — CHANGES_REQUESTED** [review]
Haven't properly tested out yet but just wanted to do a first pass on the code. Looks good in general but I'm worried about the diff in package-lock. Are you sure that's just the addition of the hardhat package? Looks like it might have update a bunch of other packages too and I'm wondering why.

File: package.json [hunk 1]
```diff
@@ -16,7 +16,8 @@
     "prepare": "node ./prepare.js",
     "tailwind:build": "tailwind-config-viewer export ./tailwind-static",
     "tailwind:view": "tailwind-config-viewer -o",
-    "generate:voting-gauges": "ts-node -O '{\"module\": \"commonjs\"}' ./src/lib/scripts/voting-gauge.generator.ts"
+    "generate:voting-gauges": "ts-node -O '{\"module\": \"commonjs\"}' ./src/lib/scripts/voting-gauge.generator.ts",
+    "node": "npx hardhat node --fork $(grep ALCHEMY_URL .env | cut -d '=' -f2) --fork-block-number 14828550"
```
[the remark this answers is not in the document]
Comment (jareknowotka) [reply package.json:20]: good idea

**Comment by garethfuller** [comment]
@gmbronco Maybe instead of adding hardhat to the frontend repo we can just assume the dev has a local node running?

File: .env.development [hunk 2]
```diff
@@ -1,7 +1,7 @@
 VUE_APP_ENV=development
 VUE_APP_DOMAIN=localhost:8080
 VUE_APP_IPFS_NODE=cloudflare-ipfs.com
-VUE_APP_NETWORK=42
+VUE_APP_NETWORK=31337
```
[the remark this answers is not in the document]
Comment (jareknowotka) [reply .env.development:4]: updating to 1, actually it should stay like that for a forked node as well. the only place where we will need to know if we are forking is list of tokens to let the page load without bloated RPC calls.

File: src/lib/balancer.sdk.ts [hunk 3]
```diff
@@ -17,7 +17,26 @@ const network = ((): Network => {
   }
 })();
 
-export const balancer = new BalancerSDK({
-  network,
-  rpcUrl: configService.rpc
-});
+const sdkConfig = ((): BalancerSdkConfig => {
```
[the remark this answers is not in the document]
Comment (gmbronco) [reply src/lib/balancer.sdk.ts]: fixed

File: src/lib/config/hardhat.json [hunk 4]
```diff
@@ -0,0 +1,82 @@
+{
```
Comment (garethfuller) [inline src/lib/config/hardhat.json:1]: Should we name this file something different? Maybe local.json?

File: src/fixtures/listed.tokenlist.json [hunk 5]
```diff
@@ -0,0 +1,154 @@
+{
```
Comment (garethfuller) [inline src/fixtures/listed.tokenlist.json:1]: I'd rather this fixtures folder wasn't as high level, e.g. right under /src. Maybe best to nest with /src/lib for now?

File: src/lib/scripts/tokenlists.generator.ts [hunk 6]
```diff
@@ -11,8 +12,20 @@ if (process.env.NODE_ENV === 'development') {
...
+  let tokenlists: TokenListMap;
+  // Handle special case where we are using local test node.
+  // If we use the whole list of ~100 tokens, node will most likely crash
+  if (process.env.VUE_APP_NETWORK == '31337') {
+    tokenlists = JSON.parse(
+      fs.readFileSync(
+        path.resolve(__dirname, '../../fixtures/listed.tokenlist.json'),
+        'utf8'
+      )
+    );
```
Comment (garethfuller) [inline src/lib/scripts/tokenlists.generator.ts:24]: Can you not just use a require statement here?
```suggestion
    tokenlists = require('../../fixtures/listed.tokenlist.json');
```

**Review by garethfuller — CHANGES_REQUESTED** [review]

**Comment by gmbronco** [comment]
> Haven't properly tested out yet but just wanted to do a first pass on the code. Looks good in general but I'm worried about the diff in package-lock. Are you sure that's just the addition of the hardhat package? Looks like it might have update a bunch of other packages too and I'm wondering why.

i rebased from develop, so it's not the case anymore

File: .env.development (hunk 2 above)
[the remark this answers is not in the document]
Comment (gmbronco) [reply .env.development]: synced with develop to mainnet here as well. i added some info in README how to setup a forked node

File: src/providers/tokens.provider.ts [hunk 7]
```diff
@@ -251,7 +250,7 @@ export default {
      */
     function mapTokenListTokens(tokenLists: TokenList[]): TokenInfoMap {
       const tokensMap = {};
-      const tokens = [...tokenLists, balTokenList]
+      const tokens = [...tokenLists]
```
[the remark this answers is not in the document]
Comment (gmbronco) [reply src/providers/tokens.provider.ts]: yes, rebased

File: src/fixtures/listed.tokenlist.json (hunk 5 above)
Comment (gmbronco) [reply src/fixtures/listed.tokenlist.json:1]: how about top level src/test folder? there is `__mocks__` which could go to src/test/mocks as well?

Comment (garethfuller) [reply src/fixtures/listed.tokenlist.json:1]: Yeah, I like that but I'm worried that there is some auto-magic going on with that mocks folder. Could you try moving it and run the tests to see if they still pass?

**Comment by timjrobinson** [comment]
We reset the `develop` branch to master which confused Github, so I've done a rebase of this branch onto the new develop and force pushed it back. All code changes should be the same.

File: README.md [hunk 8]
```diff
@@ -53,8 +53,33 @@ To change your local development app network, update the network key for
...
+For local development purposes you may want to use a test node with a forked mainnet state.
+
+If you keep you RPC_URL in local .env
+```
+export RPC_URL=$(grep ALCHEMY_URL .env | cut -d '=' -f2)
+```
+
+Hardhat:
+```
+npm run node
```
Comment (sssmi) [inline README.md:70]: Maybe add a note that this command should be run from `balancer-sdk`.

Also listing available network ids on the readme could be helpful.

**Comment by garethfuller** [comment]
Closing in favour of #2082, something seemed to break the diff after the develop reset.

**Pull request closed without merging**